README
****************************************************
*                                                  *
*    GtkSu ©K.D.Hedger 2013-2019                   *
*                                                  *
*    keithdhedger@gmail.com                        *
*                                                  *
*                                                  *
****************************************************


Gtksu - what is it?

Simple its a simple replacemnet for gksu/ktsuss etc that alows you to run a program with diferrent priviliges ( root etc ).

Although there are other programs of theis type most if not all rely on polkit or have a large number of dependencies, apps like ktsuss which this app was inspired by ( avaiable here:http://ktsuss.googlecode.com/files/ktsuss-2.tar.gz ), some like ktsuss seem to have been abandoned ( last commit was 2011 ).

Also it seems that whenever polkit ( which is a nightmare! ) is updated these programs break, so this is the result there are few dependencies a reasonable recent gtk and glib and thats it.

Installation

Just run ./autogen.sh or ./configure from this folder then make and sudo make install.
To install to ( say ) /usr instead of the default /usr/local, add --prefix=/usr, --enable-ktsuss-link will as it says link ktsuss to gtksu so that some programs that hard code the files think that ktsuss is being used ( spacefm ).

You can also choose to build a QT5 version by adding --enable-qt5 to the configure command, the default is to build a gtk GUI.
To install Qt5 go to here:
http://qt-project.org/downloads

Download the required file and then do this:
http://qt-project.org/wiki/Install_Qt_5_on_Ubuntu

If you have installed in /opt/... which is the default for the installer to compile with the --enable-qt5 flag add this line:
/opt/Qt5.2.1/5.2.1/gcc_64/lib
to /etc/ld.so.conf
and do this:
PKG_CONFIG_PATH=/opt/Qt5.2.1/5.2.1/gcc_64/lib/pkgconfig ./configure --prefix=/usr --enable-ktsuss-link --enable-qt5

If you have installed Qt5 elsewhere just adjust the /etc/ld.so.conf and PKG_CONFIG_PATH to suit.

If your distro disables the root account ( I'm looking at you Ubuntu! ) you may have to set a root password like so:
Open a Terminal and type:
sudo su
passwd

Enter a new password for root twice you may also have to run these two commands whilst logged in as root:

pwconv
grpconv

Finally logout
exit

You can now type roots password in to the GtkSu gui and run commands as root.

Usage:

gtksu [OPTION] [--] <command>
Run a command as another user
-u, --user USER			Runs the command as the given user
-m, --message MESG		Change default message in ktsuss window

Simple usage:
gtksu terminal.

If you need to add options use -- like so:
gtksu -u root -- terminal --hold -x ls -AF /

This programme has the potential to hose your system if used unwisely, BE WARNED!
May present a security risk so use at your own risk.

Bugs etc to the above email address.


ChangeLog
0.1.6
New email address.
Removed redundant scripts folder.
Updated manpage.
Removed forced CFLAGS and CXXFLAGS.
Added to Github:https://github.com/KeithDHedger/GtkSu

0.1.5
Updated licence info in files.
Fixed some distros not setting some users environment variables.

0.1.4
Command line to gtksu can now be quoted or not eg to open the file "/tmp/dat 1" all these are the same:
gtksu -- leafpad /tmp/dat\ 1
gtksu -- "leafpad /tmp/dat\ 1"
gtksu -- 'leafpad /tmp/dat\ 1'
gtksu -- leafpad "/tmp/dat 1"
gtksu -- 'leafpad \"/tmp/dat 1\"'

0.1.3
Added submitted patch to fix minor bug when using --user switch, thanks to Ludger Krämer for that.

0.1.2
Removed unnecessary .desktop and png file.
Minor update to man page.

0.1.1
Swapped butons in gtk.
Fixed not displaying default message.
Fixed not displaying alternate message.
Added option to build with QT5 instead of Gtk ( default ).

0.1.0
Updated manpage.
Load more error checking.
CD to users home or "/" before running command.
Only roots PATH now includes /sbin and /usr/sbin.
Cleaned environment before switching user.
Returns exit status of su'ed program.
Aborts if environment can't be cleaned or permissions can't be dropped.
Permissions in suwrap program only elevated when needed.
Fixed make asking for admin password.

0.0.8
Fixed install of zipped manpage.
Fixed wrong version number being picked up.

0.0.7 
Added install instructions for debian/fedora users including dependenxies.
Added 'configure' file etc so now no longer depends on autotools.

0.0.6
Fixed a bug where '--' used to mark end of options used when no options.
Removed debugging messages.

0.0.5
First release into the wild.



GtkSu/src/internet.h
/*
 *
 * ©K. D. Hedger. Tue 28 Jun 10:44:42 BST 2016 keithdhedger@gmail.com
 
 * Projects is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * at your option) any later version.

 * Projects is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with Projects.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef _INTERNET_
#define _INTERNET_

#define MYEMAIL "keithdhedger@gmail.com"
#define MYWEBSITE "http://khapplications.freeddns.org/index.html"
#define GLOBALWEBSITE "http://khapplications.freeddns.org"
#define ASPELLPAGE GLOBALWEBSITE "/pages/aspellgui/aspelgui.html"
#define MANPAGEPAGE GLOBALWEBSITE "/pages/manpageeditor/help.html"
#define CLIPVIEW GLOBALWEBSITE "/pages/clipviewer/clip.html"
#define THEMEMANAGER GLOBALWEBSITE "/pages/apps.html#themeed"
#define COMPMANAGER GLOBALWEBSITE "/pages/apps.html#xfcecomp"
#define KKEDITPAGE "http://kkedit.freeddns.org"
#define KKTERMINALPAGE "http://khapplications.freeddns.org/pages/kkterminal/kkterminal.html"

#define COPYRITE "Copyright \xc2\xa9 2013-2019 K.D.Hedger"

#ifndef _USEQT5_
extern const char	*moreapps[];
extern const char	*authors[];
#else
extern const char*	authors;
#endif
#endif


GtkSu/MakeSuWrap/suwrap.cpp
/*
 *
 * ©K. D. Hedger. Sun 27 Sep 20:11:17 BST 2015 keithdhedger@gmail.com

 * This file (suwrap.cpp) is part of GtkSu.

 * GtkSu is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * at your option) any later version.

 * GtkSu is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with GtkSu.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Parts of this code are from udevil.c available here:
 * https://nodeload.github.com/IgnorantGuru/udevil/zip/master
 * Mucked about by me :)
 *
 * And libgksu.c available here:
 * http://people.debian.org/~kov/gksu/libgksu-2.0.12.tar.gz
 * Mucked about by me :)
 *
 * All code is suppled 'as is' use it at your own peril!
 * Released under GPLv3
 *
 */

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <errno.h>
#include <sys/types.h>
#include <grp.h>
#include <sys/wait.h> 

#include "config.h"

#include <glib.h>

#include <pwd.h>
#include <shadow.h>
#include <crypt.h>

#define	NOSHADOWUSER 255
#define CANTDROPPRIVS 254
#define NOXAUTHBIN 253
#define NOCLEANENV 252
#define CANTCHDIR 251
#define BADPASSWD 250
#define CANTMAKETMPDIR 249
#define CANTMAKEXAUTHFILE 248

static int		orig_ngroups=-1;
static gid_t	orig_groups[NGROUPS_MAX];
static gid_t	orig_rgid=-1;
static uid_t	orig_ruid=-1;
static gid_t	orig_egid=-1;
static uid_t	orig_euid=-1;

bool			firstRun=true;
int				retFromApp=-1;
char*			userHome;
char*			userName;
char*			userShell;
char*			userDisplay;
char*			userXAuth;
char*			userTz;
char*			userLang;
char*			userLcAll;
char*			userLCCol;
char*			userLCCType;

struct passwd*	pwdata;
extern char**	environ;
char*			xauthDir=NULL;
char*			xauthFile=NULL;

void setEnvTest(const char *env,const char *envvar)
{
	if(envvar!=NULL)
		setenv(env,envvar,1);
}

void drop_privileges(int permanent)
{
	if (geteuid()!=0)
		return;

	if (firstRun==true)
		{
			firstRun=false;
			orig_euid=geteuid();
			orig_egid=getegid();
			orig_ruid=getuid();
			orig_rgid=getgid();
			orig_ngroups=getgroups(NGROUPS_MAX,orig_groups);
		}

// drop groups
/* If root privileges are to be dropped, be sure to pare down the ancillary
* groups for the process before doing anything else because the setgroups(  )
* system call requires root privileges.  Drop ancillary groups regardless of
* whether privileges are being dropped temporarily or permanently.
*/
	gid_t newgid=orig_rgid;
	setgroups(1,&newgid);

	if (setregid(permanent ? newgid : -1,newgid)==-1) goto _drop_abort;

    // drop user
	if (setreuid((permanent ? orig_ruid : -1),orig_ruid)==-1) goto _drop_abort;

    // verify if not originally root
	if (orig_ruid!=0)
		{
			if (permanent)
				{
					if ((setegid(0)!=-1) || (getegid()!=newgid))
						goto _drop_abort;
					if ((seteuid(0)!=-1) || (geteuid()!=orig_ruid))
						goto _drop_abort;
				}
			else
				{
					if (getegid()!=newgid)
						goto _drop_abort;
					if (geteuid()!=orig_ruid)
						goto _drop_abort;
				}
		}

	return;

_drop_abort:
	fprintf(stderr,"gtksuwrap: error 1: unable to drop priviledges - please report this problem\n");
    exit(CANTDROPPRIVS);
}

void restore_privileges(void)
{
	if (orig_euid!=0)
		return;

	seteuid(0);
	setegid(orig_egid);
	setgroups(orig_ngroups,orig_groups);
}

int checkPasswd(char* username,char* hashedpass)
{
	spwd*	shadow_entry=NULL;

	restore_privileges();
		shadow_entry=getspnam(username);
	drop_privileges(0);
	if(shadow_entry==NULL)
		return(NOSHADOWUSER);
	else
		return(strcmp(hashedpass,shadow_entry->sp_pwdp));
}

int sendHashBack(char* username)
{
	spwd*	shadow_entry=NULL;

	restore_privileges();
		shadow_entry=getspnam(username);
	drop_privileges(0);
	if(shadow_entry==NULL)
		{
			printf("XXX\n");
			return(NOSHADOWUSER);
		}
	else
		{
			printf("%s\n",shadow_entry->sp_pwdp);
			return(0);
		}
}

void keepEnvs(int theuid)
{
	pwdata=getpwuid(theuid);
	userHome=pwdata->pw_dir;
	userName=pwdata->pw_name;
	userShell=pwdata->pw_shell;
	userDisplay=getenv("DISPLAY");
	userXAuth=getenv("XAUTHORITY");
	userTz=getenv("TZ");
	userLang=getenv("LANG");
	userLcAll=getenv("LC_ALL");
	userLCCol=getenv("LC_COLLATE");
	userLCCType=getenv("LC_CTYPE");
}

void makeXauthFile(void)
{
	char*		command;
	gchar		tname[]="/tmp/GtkSu-XXXXXX";
	const char*	xauthBinPath=NULL;

	if(g_file_test("/usr/bin/xauth",G_FILE_TEST_IS_EXECUTABLE))
		xauthBinPath="/usr/bin/xauth";
	if(g_file_test("/usr/X11R6/bin/xauth",G_FILE_TEST_IS_EXECUTABLE))
		xauthBinPath="/usr/X11R6/bin/xauth";

	if(xauthBinPath==NULL)
		{
			fprintf(stderr,"Can't find xauth binary in /usr/bin/ or /usr/X11R6/bin/xauth\n");
			exit(NOXAUTHBIN);
		}

	xauthDir=strdup(mkdtemp(tname));
	if(xauthDir==NULL)
		exit(CANTMAKETMPDIR);

	asprintf(&xauthFile,"%s/.Xauthority",xauthDir);
	asprintf(&command,"chmod 777 %s",xauthDir);
	system(command);
	free(command);
	asprintf(&command,"%s extract - %s > %s",xauthBinPath,userDisplay,xauthFile);
	system(command);
	free(command);
}

void cleanEnv(int theuid,bool createxauth)
{
	keepEnvs(theuid);

#ifdef _GOTCLEARENV_
	if(clearenv()!=0)
		{
			fprintf(stderr,"Can't clean environment, aborting ...");
			exit(NOCLEANENV);
		}
#else
	environ=NULL;
#endif

	if(theuid==0)
		setenv("PATH",_PATH_STDPATH,1);
	else
		setenv("PATH",_PATH_DEFPATH,1);

	setenv("IFS"," \t\n",1);
	setenv("HOME",userHome,1);
	setenv("USERNAME",userName,1);
	setenv("SHELL",userShell,1);


	setEnvTest("DISPLAY",userDisplay);
	setEnvTest("TZ",userTz);
	setEnvTest("LANG",userLang);
	setEnvTest("LC_ALL",userLcAll);
	setEnvTest("LC_COLLATE",userLCCol);
	setEnvTest("LC_CTYPE",userLCCType);
	if(chdir(userHome)!=0)
		{
			if(chdir("/")!=0)
				{
					fprintf(stderr,"Can't change PWD, aborting ...");
					exit(CANTCHDIR);
				}
		}
}

int main(int argc,char **argv)
{
	int			j;
	GString*	str;
	int			theuid;
	int			retval;

	drop_privileges(0);
	cleanEnv(geteuid(),false);

	theuid=atoi(argv[1]);
	str=g_string_new(NULL);

	if(strncmp(argv[1],"gethash",7)==0)
		{	
			retval=sendHashBack(argv[2]);
			return(retval);
		}

	if(checkPasswd(argv[2],argv[3])==0)
		{
			for(j=4;j<argc;j++)
				{
					if(j==4)
						g_string_append_printf(str," %s",argv[j]);
					else
						g_string_append_printf(str," \"%s\"",argv[j]);
				}
			makeXauthFile();
			restore_privileges();
				setresuid(theuid,theuid,theuid);
				cleanEnv(geteuid(),false);
				setEnvTest("XAUTHORITY",xauthFile);
				retFromApp=system(str->str);
			drop_privileges(0);
			unlink(xauthFile);
			rmdir(xauthDir);
			g_string_free(str,true);
			free(xauthDir);
			free(xauthFile);

			return(WEXITSTATUS(retFromApp));
		}
	else
		{
			return(BADPASSWD);
		}
}


GtkSu/src/main.cpp
/*
 *
 * ©K. D. Hedger. Sun 27 Sep 20:10:42 BST 2015 keithdhedger@gmail.com

 * This file (main.cpp) is part of GtkSu.

 * GtkSu is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * at your option) any later version.

 * GtkSu is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with GtkSu.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <errno.h>
 
#include <pwd.h>
#include <shadow.h>
#include <crypt.h>
#include <getopt.h>

#include "config.h"
#include "internet.h"

#define		NOSHADOWUSER -1
#define		GTKPADDING 4
char*		user;
char*		passwd;

#ifndef _USEQT5_
#include <gtk/gtk.h>
GtkWidget*	window=NULL;
GtkWidget*	nameEntry=NULL;
GtkWidget*	passEntry=NULL;

#else
#include <glib.h>

#include <QtWidgets>
#include <QObject>
QApplication*	holdapp;
QWidget*		mainWindow;
#endif


char**		gargv;
GString*	commandStr=g_string_new(NULL);

char*		whereFrom;
char*		hashedPass=NULL;
char*		userName=NULL;
char*		bodyMessage=NULL;

GString*	runOptions=g_string_new(NULL);

int			returnValFromApp=-1;

void shutItDown(gpointer x,gpointer y)
{
#ifdef _USEQT5_
	return(holdapp->exit(WEXITSTATUS(returnValFromApp)));
#else
	gtk_main_quit();
#endif
}

struct option long_options[]=
	{
		{"version",0,0,'v'},
		{"user",1,0,'u'},
		{"message",1,0,'m'},
		{"help",0,0,'h'},
		{0, 0, 0, 0}
	};

int runAsUser(int theuid,char*user,char* hashedpass)
{
	GString*	str=g_string_new(NULL);

	g_string_append_printf(str,"%s/gtksuwrap %i '%s' '%s' %s",whereFrom,theuid,user,hashedpass,commandStr->str);
#ifndef _USEQT5_
	gtk_widget_hide(window);
	while(gtk_events_pending())
		gtk_main_iteration_do(false);
#else
	mainWindow->hide();
#endif
	returnValFromApp=system(str->str);

	g_string_free(str,true);

	return(0);
}

void doErrorMessage(const char* message,const char* data,const char* secondmessage)
{
	char*	vmessage;

	asprintf(&vmessage,"%s %s\n%s",message,data,secondmessage);
#ifndef _USEQT5_
	GtkWidget*	dialog;
	dialog=gtk_message_dialog_new((GtkWindow*)window,GTK_DIALOG_MODAL,GTK_MESSAGE_ERROR,GTK_BUTTONS_CLOSE,"%s %s\n",message,data);
	gtk_message_dialog_format_secondary_text((GtkMessageDialog*)dialog,"%s\n",secondmessage);
	gtk_dialog_run((GtkDialog*)dialog);
	gtk_widget_destroy(dialog);
#else
	QMessageBox		msgBox;

	msgBox.setText(vmessage);
	msgBox.setIcon(QMessageBox::Critical);
	msgBox.exec();

#endif
	free(vmessage);
}

void printHelp(void)
{
	printf("GtkSu Version %s \nCopyright K.D.Hedger 2013-2019, %s\n",VERSION,MYEMAIL);
	printf("Usage: gtksu [OPTION] [--] <command>\n");
	printf("Run a command as another user\n");
	printf("-u, --user USER			Runs the command as the given user\n");
	printf("-m, --message MESG		Change default message in ktsuss window\n");

}

void getPath(void)
{
	char	arg1[32];
	char	exepath[PATH_MAX+1]={0};

	sprintf(arg1,"/proc/%d/exe",getpid());
	readlink(arg1,exepath,1024);
	whereFrom=g_path_get_dirname(exepath);
}

#ifdef _USEQT5_
QLineEdit*		nameBox;
QLineEdit*		passBox;

#endif

void doGoForIt(void)
{
	FILE*		fp;
	char		buffer[256];
	int			retval;
	char*		command;
	char*		resulthash=NULL;
	int			uid;
	int			itworked;

	sprintf(buffer,"id -u %s 2>/dev/null",user);
	fp=popen(buffer, "r");
	fgets(buffer,64,fp);
	buffer[strlen(buffer)-1]=0;
	retval=pclose(fp);
	if(retval==0)
		{
			uid=atoi(buffer);
			errno=0;
			buffer[0]=0;
			asprintf(&command,"%s/gtksuwrap gethash %s",whereFrom,user);
			fp=popen(command,"r");
			g_free(command);
			if(fp!=NULL)
				{
					fgets(buffer,255,fp);
					buffer[strlen(buffer)-1]=0;
					pclose(fp);
					asprintf(&hashedPass,"%s",buffer);
					resulthash=crypt(passwd,hashedPass);

					if((resulthash!=NULL) && (strcmp(hashedPass,resulthash)==0))
						{
							itworked=runAsUser(uid,user,resulthash);
									if(itworked==0)
										shutItDown(NULL,NULL);
						}
					else
						{
							doErrorMessage("Could not run ",commandStr->str,"Username and/or Password incorrect");
						}
				}
			return;
		}
	else
		doErrorMessage("Unknown User ",user,"");

}

void doApply(void)
{
#ifdef _USEQT5_
	user=strdup(nameBox->text().toUtf8().constData());
	passwd=strdup(passBox->text().toUtf8().constData());
#else
	user=strdup(gtk_entry_get_text((GtkEntry*)nameEntry));
	passwd=strdup(gtk_entry_get_text((GtkEntry*)passEntry));
#endif
	doGoForIt();
}

int main(int argc,char **argv)
{
	int c;
	int option_index=0;

	while (1)
		{
			option_index=0;
			c=getopt_long_only(argc,argv,"u:m:v?h",long_options,&option_index);

			if (c==-1)
				break;

			switch (c)
				{
					case '?':
					case 'h':
						printHelp();
						return 0;
						break;
			
					case 'u':
						userName=optarg;
						break;

					case 'v':
						printf("GtkSu Version %s \n%s, %s\n",VERSION,COPYRITE,MYEMAIL);
						return 0;
						break;

					case 'm':
						bodyMessage=optarg;
						break;

					default:
						printf ("?? Unknown argument ??\n");
						break;
			}
		}

	gargv=argv;
	getPath();

	for(int j=optind;j<argc;j++)
		g_string_append_printf(commandStr," \"%s\"",argv[j]);

#ifdef _USEQT5_
	QApplication	app(argc, argv);

	QVBoxLayout*	vlayout=new QVBoxLayout;
	QHBoxLayout*	hlayout;
	QPushButton*	cancelButton=new QPushButton("&Cancel");
	QPushButton*	okButton=new QPushButton("&Apply");
	QWidget*		hbox;
	QLabel*			label;

	holdapp=&app;
	mainWindow=new QWidget;
	nameBox=new QLineEdit;
	passBox=new QLineEdit;
	nameBox->setText("root");

	if(userName!=NULL)
		{
			nameBox->setText(userName);
			nameBox->hide();
		}
	else
		nameBox->setText("root");

	if(bodyMessage!=NULL)
			label=new QLabel(bodyMessage);
	else
		label=new QLabel("Please enter the desired username and password:");

	hlayout=new QHBoxLayout;
	hlayout->setContentsMargins(0,0,0,0);
	hbox=new QWidget;
	hbox->setLayout(hlayout);
	label->setAlignment(Qt::AlignCenter);
	hlayout->addWidget(label);
	vlayout->addWidget(hbox);

	hlayout=new QHBoxLayout;
	hlayout->setContentsMargins(0,0,0,0);
	hbox=new QWidget;
	hbox->setLayout(hlayout);
	if(userName==NULL)
		{
			hlayout->addWidget(new QLabel("User Name"),Qt::AlignLeft);
			hlayout->addWidget(nameBox,Qt::AlignRight);
			vlayout->addWidget(hbox);
		}

	hlayout=new QHBoxLayout;
	hlayout->setContentsMargins(0,0,0,0);
	hbox=new QWidget;
	hbox->setLayout(hlayout);
	hlayout->addWidget(new QLabel("Password"),Qt::AlignLeft);
	hlayout->addWidget(passBox,Qt::AlignRight);
	passBox->setEchoMode(QLineEdit::Password);
	vlayout->addWidget(hbox);

	hbox=new QWidget;
	hlayout=new QHBoxLayout;
	hlayout->setContentsMargins(0,0,0,0);
	hbox->setLayout(hlayout);
	hlayout->addWidget(cancelButton);
	hlayout->addStretch(0);
	hlayout->addWidget(okButton);

	vlayout->addWidget(hbox);

	QObject::connect(cancelButton,SIGNAL(clicked()),qApp,SLOT(quit()));
	QObject::connect(okButton,&QPushButton::clicked,doApply );
	QObject::connect(passBox,&QLineEdit::returnPressed,doApply );

	mainWindow->setLayout(vlayout);

	passBox->setFocus();
	mainWindow->show();

	app.exec();

#else
	GtkWidget*	vbox;
	GtkWidget*	hbox;
	GtkWidget*	buttonok;
	GtkWidget*	button;

	gtk_init(&argc,&argv);

	window=gtk_window_new(GTK_WINDOW_TOPLEVEL);
	gtk_window_set_title((GtkWindow*)window,"GtkSu");
	g_signal_connect(G_OBJECT(window),"delete-event",G_CALLBACK(shutItDown),NULL);

	vbox=gtk_vbox_new(false,0);
	nameEntry=gtk_entry_new();
	g_signal_connect_after(G_OBJECT(nameEntry),"activate",G_CALLBACK(doApply),NULL);

	if(bodyMessage!=NULL)
		gtk_box_pack_start(GTK_BOX(vbox),gtk_label_new(bodyMessage),false,true,GTKPADDING);
	else
		gtk_box_pack_start(GTK_BOX(vbox),gtk_label_new("Please enter the desired username and password:"),false,true,GTKPADDING);

	if(userName!=NULL)
		{
			gtk_entry_set_text((GtkEntry*)nameEntry,userName);
			gtk_widget_hide(nameEntry);
		}
	else
		{
			hbox=gtk_hbox_new(false,0);
			gtk_entry_set_text((GtkEntry*)nameEntry,"root");
			gtk_box_pack_start(GTK_BOX(hbox),gtk_label_new("User Name\t"),false,true,0);
			gtk_box_pack_start(GTK_BOX(hbox),nameEntry,true,true,0);
			gtk_box_pack_start(GTK_BOX(vbox),hbox,true,true,GTKPADDING);
		}

	passEntry=gtk_entry_new();
	gtk_entry_set_visibility((GtkEntry*)passEntry,false);
	g_signal_connect_after(G_OBJECT(passEntry),"activate",G_CALLBACK(doApply),NULL);

	hbox=gtk_hbox_new(false,0);
	gtk_box_pack_start(GTK_BOX(hbox),gtk_label_new("Password\t"),false,true,0);
	gtk_box_pack_start(GTK_BOX(hbox),passEntry,true,true,0);
	gtk_box_pack_start(GTK_BOX(vbox),hbox,true,true,GTKPADDING);

	hbox=gtk_hbutton_box_new();
	gtk_button_box_set_layout((GtkButtonBox*)hbox,GTK_BUTTONBOX_SPREAD);
	
	button=gtk_button_new_from_stock(GTK_STOCK_CANCEL);
	gtk_container_add(GTK_CONTAINER(hbox),button);
	g_signal_connect(G_OBJECT(button),"clicked",G_CALLBACK(shutItDown),NULL);

	gtk_box_pack_start(GTK_BOX(hbox),gtk_label_new(""),true,true,0);

	buttonok=gtk_button_new_from_stock(GTK_STOCK_APPLY);
	gtk_container_add(GTK_CONTAINER(hbox),buttonok);
	g_signal_connect(G_OBJECT(buttonok),"clicked",G_CALLBACK(doApply),NULL);

	gtk_box_pack_start(GTK_BOX(vbox),hbox,true,true,GTKPADDING);
	gtk_container_add(GTK_CONTAINER(window),vbox);

	gtk_widget_show_all(window);

	if(userName!=NULL)
		gtk_widget_hide(nameEntry);

	gtk_widget_grab_focus(passEntry);
	gtk_main();
#endif

	return(WEXITSTATUS(returnValFromApp));
}



